site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md

```diff
@@ -0,0 +1,437 @@
+
+This has some practical consequences:
+* You can revert a revert: it's just like any other commit.
+* You can push your branch, containing the revert, and submit a pull request to review your changes.
+* If you're working out of a shared branch, your peers will pull the revert like any other commit. This is usually the right way to "undo changes" in a shared branch, vs. using a command that rewrites history and requires a force push instead.
+
+Of course, if you've ever done this before, you know that I've kind of cheated today: my revert was clean, in that there were no merge conflicts. Sometimes, especially if some time has passed, other changes might have touched the same lines, and the revert will require resolving those changes. Hold that thought, though, since I'll work through an example of resolving merge conflicts with the next command. 
+Remember, reverts are just new commits!
+
+### git cherry-pick
+
+Next up: cherry picking! The general idea of cherry picking -- not specific to git -- is to pick and choose just what you want. You can imagine yourself standing in front of an actual cherry tree, and saying, "I'll take that one, and that one, and that one," and ignoring the rest.
+
+In git, the idea is similar. Given one or more commits -- anywhere in the tree -- apply those same changes to my current ref. Why would you want to do that?
+
+At Mattermost, master is our development branch. Every two months we [release a new version](https://developers.mattermost.com/contribute/getting-started/branching/) of Mattermost that branches off master. Our release team begins to qualify this release, inevitably finding issues. We first fix the bugs in master, and then cherry-pick those changes back to the release branch. Of course, there are lots of different ways of using Git to handle releases, but this is the strategy we use at Mattermost.
+
+![image](/blog/2020-03-16-advanced-git-tbilisi-free-university/branching-overview.png)
+
+Let's look at the most recent such feature release: v5.20.
+
+```sh
+# Checkout the mattermost-server repository (if not already done)
+git checkout https://github.com/mattermost/mattermost-server.git
+cd mattermost-server
+
+# Examine the combined history of release-5.20 and master
+git log --graph --oneline release-5.20 master
+```
+
+Observe that master has advanced in the weeks since we released v5.20. When we first released v5.20.0, a community member tried installing it and promptly [reported](https://community.mattermost.com/core/pl/877mkcc967ytjpgcgpbo6bi7ty) that his server was crashing! A bit of debugging on a Saturday, and I realized that we had introduced a regression into a code path we didn't often use ourselves. I filed [MM-21619](https://mattermost.atlassian.net/browse/MM-22619) and got to work on the fix. Let's find it in the master branch.
```

Comment:
```suggestion
Observe that `master` has advanced in the weeks since we released v5.20. When we first released v5.20.0, a community member tried installing it and promptly [reported](https://community.mattermost.com/core/pl/877mkcc967ytjpgcgpbo6bi7ty) that his server was crashing! A bit of debugging on a Saturday, and I realized that we had introduced a regression into a code path we didn't often use ourselves. I filed [MM-21619](https://mattermost.atlassian.net/browse/MM-22619) and got to work on the fix. Let's find it in the `master` branch.
```